Week2/LESSONPLAN.md

```diff
@@ -0,0 +1,483 @@
+# Lesson Plan Databases Week 2
+
+The lesson plan is primarily written for teachers so that they can
+use examples and anecdotes from this document in conjunction with the README
+and explain the concepts better in the class.
+
+## Topics
+
+0. Async nature of MySQL-Nodejs interaction
+1. Keys (unique, primary)
+2. Relationships (1-M with foreign keys, M-M with composite keys)
+3. Joins (comma (cross), implicit inner, self, left, right)
+4. More SQL clauses (group by, having, distinct and Aggregate functions)
+5. Domain modeling (ERD)
+
+
+### 0. Async nature of MySQL-Nodejs interaction
+
+#### Explanation
+The nature of database queries is asynchronous.
+Some queries can take long time to execute.
+When our JavaScript MySQL client is sending the queries to the MySQL server,
+it may not want to block until the answer is returned.
+However if the JavaScript MySQL client is sending multiple queries such that
+the second query (for example insert) depends on the first query (for example create),
+then it must wait until the execution of the first query is successful.
+To ensure smooth interaction with the MySQL server, promises can be used in conjunction
+with the await() method.
+
+#### Example(s)
+Demonstrate with four programs at [this repository](https://github.com/unmeshvrije/database_examples)
+how
+1. Program `1-db-naive.js` fails because the connection is closed prematurely.
+2. Program `2-db-callback.js` solves the problem but looks ugly because of the callback-hell.
+3. Program `3-db-promise.js` uses the promise chaining to make it better.
+about building those promises
+4. Program `4-db-await.js` uses promisify() and await() to make it the best.
+
+#### Exercise
+The program called `async-create-insert.js` can be found in `Week2` folder.
+Add a select query to that program using await and promisify.
+
+#### Essence
+> async keyword : to create asynchronous function and ensure they return promise without having to worry
+
+> await : to call a function returning promise without having to call .then() over that promise
+
+> promisify() : to convert a callback based function to a promise based one.
+
+
+### 1. Key (Constraints)
+#### Explanation
+1. A column can be declared as the UNIQUE column. Such a column has UNIQUE values.
+It can also have NULL values. Thus, two rows can have same NULL value in the column that
+is declared as UNIQUE (In other words, this is a UNIQUE CONSTRAINT on that column).
+2. A column can be declared as the PRIMARY KEY column. Such a column has UNIQUE values too.
+They cannot be NULL values. Thus two rows can NEVER have same values in the column
+that is declared as PRIMARY KEY (In other words, this is a PRIMARY KEY CONSTRAINT on that column).
```

Comment:
I would have something about the Sequence and the auto-increment because it's really useful

Reply:
It's an excellent suggestion! Thanks :)